[FEATURE] Next.js frontend setup

Mosaab-Emam opened the pull request:
# Pull Request

## 📋 Description

Created /frontend with a base frontend application. It uses React & Next.js. It's fully integrated with the /backend auth system, including sign up, sign in, sign out, password reset request, and password update functionality. Features automatic token refresh, server-side route protection via Next.js middleware, and secure httpOnly cookie-based session management. Has minimal UI until the app styling is finalized.

## 🔗 Related Issue

- Related #42. Should consider [frontend README.md](https://github.com/CivicTechWR/go-train-group-pass/blob/c49d8d7ec91104d6d5bab6f453e836b7ab01a18a/frontend/README.md)
- Closes #26 

## 🧪 Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update

## 🎯 Changes Made

Created /frontend that contains the base frontend for the app using React & Next.js. It's features are documented in [frontend README.md](https://github.com/CivicTechWR/go-train-group-pass/blob/c49d8d7ec91104d6d5bab6f453e836b7ab01a18a/frontend/README.md), summarized as follows:
 
- [x] Implemented React Context API for global authentication state management
- [x] Created authentication forms (SignIn, SignUp, PasswordResetRequest, PasswordUpdate)
- [x] Built Next.js API routes that proxy to backend with Zod validation
- [x] Implemented Next.js middleware for server-side route protection
- [x] Added automatic token refresh in both API client and middleware
- [x] Created protected and public pages with proper routing

## 🧪 Testing

- [ ] Unit tests pass
- [ ] Integration tests pass
- [ ] E2E tests pass
- [x] Manual testing completed
- [ ] Accessibility testing completed
- [ ] Performance testing completed

## 📱 Screenshots/Videos

<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 50 05 PM" src="https://github.com/user-attachments/assets/a7990bac-8721-4ce4-b561-f44d9b818e9d" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 50 11 PM" src="https://github.com/user-attachments/assets/d365ec1c-66a3-4b81-b25f-321cda7ff230" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 52 24 PM" src="https://github.com/user-attachments/assets/9cd1ede6-985a-4cdf-9ccd-5b87a56b312e" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 52 34 PM" src="https://github.com/user-attachments/assets/bf0b0f4b-4ff9-4d72-a823-8211aa757043" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 52 45 PM" src="https://github.com/user-attachments/assets/81422286-a025-4445-a06b-c7f1bda5402a" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 53 01 PM" src="https://github.com/user-attachments/assets/d250383c-d9e3-4149-b645-e12a682813b0" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 53 10 PM" src="https://github.com/user-attachments/assets/508cb0ea-b10c-43e5-8d98-01d6f4872214" />
<img width="1366" height="868" alt="Screenshot 2025-11-16 at 1 53 37 PM" src="https://github.com/user-attachments/assets/97ac7207-ca33-43c9-8c4e-bb2c82c9d6b7" />

## 📋 Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## 🚀 Deployment Notes

- Set `NEXT_PUBLIC_API_URL` environment variable to production backend URL
- Frontend runs on port 3001 by default (configured in package.json)

## 📚 Additional Context

**Key Features:**
- **Authentication**: Sign up, sign in, sign out with email/password
- **Password Management**: Request password reset email and update password (authenticated users)
- **Route Protection**: Server-side middleware protects all routes except public ones
- **Automatic Token Refresh**: Seamless token refresh on both client-side API calls and server-side middleware
- **Security**: httpOnly cookies, Zod validation, secure cookie settings for production
- **State Management**: React Context provides global auth state accessible throughout the app

**Architecture:**
- Next.js API routes act as proxy between frontend and NestJS backend
- All API responses validated with Zod schemas
- Tokens stored in httpOnly cookies (prevents XSS attacks)
- Middleware validates tokens before page rendering (no flash of unauthenticated content)

**Note:** Password reset from email links is not yet implemented (requires backend support for recovery tokens). The reset-password page currently works for authenticated users only. See #43 

---

**CivicTechWR Community**: Thank you for contributing to civic technology that serves the Waterloo Region community! 🚂

Commit:
restore migrations into a workable state

Commit:
make name optional, change column name in database and create base entity to add create and update dates on all objects

Commit:
add base coordination entities

Mosaab-Emam committed:
feat(frontend): init project

Mosaab-Emam committed:
chore(frontend): change the running port to 3001 to prevent conflicting with the backend

Mosaab-Emam committed:
chore(frontend): cleanup & give home page some starter identity

Mosaab-Emam committed:
fix: remove duplicate GITHUB_TOKEN line in security.yml

Mosaab-Emam committed:
feat(frontend): add api routes

Mosaab-Emam committed:
feat(frontend): Add auth primitives and /signin, /signup, /protected routes

Mosaab-Emam committed:
feat(frontend): implement password update and password reset request

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

Mosaab-Emam committed:
feat(docs): document the frontend in ./frontend/README.md

jliu1016 commented on .github/workflows/security.yml:
```diff
@@ -27,9 +27,9 @@ jobs:
       SUPABASE_SERVICE_ROLE_KEY: service-role-key
-      ENABLE_ADMIN_APIS: 'false'
+      ENABLE_ADMIN_APIS: "false"
       ADMIN_API_TOKEN: dummy-token
```
which linter is this from?

Mosaab-Emam committed:
fix: enforce project's prettier config on github workflow files

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

Commit:
initialize shadcn

Commit:
add react-hook-form

Commit:
shadify form

Commit:
merge

Commit:
recreate initial gtfs objects

Commit:
merge recreation of gtfs entities

Commit:
recreate base entities migration

Commit:
remove unnecessary if condition in auth service

Commit:
move auth into modules folder

Commit:
Merge remote-tracking branch 'origin/main' into baseEntitiesV2

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/addShadCN' into nextjs-frontend-setup

Commit:
use zod pipes for controllers and services, update user and auth services per requirements

Commit:
remove primary keys from inappropriate keys

Commit:
remove primary keys from inappropriate keys

Commit:
Merge remote-tracking branch 'origin/recreateGTFSEntities' into baseEntitiesV2

Commit:
fix entities based off wrong assumptions about gtfs trip and gtfs calendar date

Commit:
add explanation and also regenerate migration for coordination entities

Commit:
Lint

Mosaab-Emam committed:
fix: types to match new backend shapes

Mosaab-Emam committed:
fix: runtime errors

Mosaab-Emam committed:
refactor(frontend): migrate sign in and signup forms to schadcn/ui

Mosaab-Emam committed:
refactor(frontned): apply the northen lights theme

Mosaab-Emam committed:
chore: adding some responsive classes

Mosaab-Emam committed:
refactor(frontend): migrate change password and forgot password forms to schadcn/ui

Mosaab-Emam committed:
fix(frontend): small fixes

- Card header padding.
- Better hint contrast on dark mode.
- Name in greeting
- Add password reset and upate requests to auth context

Mosaab-Emam commented:
@jliu1016 Some changes were not pushed, but are now.
If you pull up to [0b8f5bf](https://github.com/CivicTechWR/go-train-group-pass/pull/44/commits/0b8f5bfd32f5bd0c0b1ca22daf16845782242ba1) you will get a working state were all the existing pages have been migrated to use Schadcn/ui components and react-hook-form for forms.

Mosaab-Emam committed:
feat(frontend): types and mock data for routes and trips

Mosaab-Emam committed:
feat: create /trips and /trips/create pages utilizing mock data

Mosaab-Emam committed:
feat: create auth respecting navbar

Mosaab-Emam commented:
Took a spin at creating pages suggested by Mohammed last week, though only mocks until related backend logic is merged into main.

<img width="565" height="815" alt="trips-page" src="https://github.com/user-attachments/assets/842958c2-e3fd-444a-bc84-bc9d513681eb" />

<br>

![create-trip-page](https://github.com/user-attachments/assets/3c462877-572c-4960-80e9-b7586b373bad)

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

jliu1016 commented:
Can we display the itineraries in more rectangular views? Similar to the look here. Will be easier to parse /scroll through vs the long cards https://www.magicpatterns.com/c/wxfdjitsvvoswg4c1nnc15

Commit:
use base entity, regenerate migration, remove unnecessary fieldName, and populate auth service interface

jliu1016 commented:
Noticed the components combine presentation and fetching/business logic. We should create as many stateless components (e.g. divide signinform to signinformcomponent which just takes props and is mostly a dummy component and a parent component which has the "useclient" on it) as possible to reduce bundle size as we really want to not be loading too much onto the users phones.
https://medium.com/@joysutradharaj/next-js-stateless-structure-building-lightweight-applications-baec1ae91884

Commit:
merge main (not working state)

Commit:
use injected repository, use feed info id

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

Mosaab-Emam committed:
feat: itineraries page, types, and mock data

Mosaab-Emam committed:
refactor: improve itineraries view

Commit:
get gtfs import to work

Commit:
make stop time import way faster by using insertMany

Commit:
remove optional from user metadata phone number

Commit:
import proper base entities

Commit:
remove infra suggestions

Commit:
check that feed is active, if inactive assume corruption and reimport

Commit:
merge

Commit:
update auth dto and schemas in shared package

Commit:
update backend configs to work with new shared package by using webpack, update auth service to use shared dtos

Commit:
update tests

Commit:
remove auth dto and schemas from backend folder

Commit:
change default of memberPresent to false, since people aren't checked in until they check in

Commit:
extend base entity on trip, fix memberPresent false on trip booking

Commit:
add cancelled enum to trip booking

Commit:
add repl

Commit:
remove migrations which were rolled up into 1 in baseEntitiesV2

Commit:
wip

remove rendundant get endpoint in auth controller

wip

Commit:
add controller to create itineraries

Commit:
remove trip schedule entity from this pr

jliu1016 commented:
IIRC isn't it that page.tsx does the fetching/state management then you have a component that does nothing?
Something like this
https://github.com/CivicTechWR/connectedkw/pull/51/files

Commit:
add trip schedule round trip

Commit:
lint

Commit:
update package.json

jliu1016 commented:
Create trip form (and we should state it as explicitly a round trip form for now) should have a drop down of "origin" and "destination" station. They also select the day. 

A get request is made to http://localhost:3000/trip-schedule/round-trip-kitchener-union?date=2025-12-09

You can see the trip-schedule.schema.ts in the shared package and use that

It will populate the KI-> Union schedule, then the Union-> KI Schedule. Each time will be selectable , when the user selects the time the formstate will store their option - including the tripCreationMetaData

Let me know if you'd like me to take this off your hands if you're floored with something else!

jliu1016 commented:
Also we are hardcoding to only allow Ki to UNion from UI rn

jliu1016 commented:
![IMG_3105](https://github.com/user-attachments/assets/102182c8-3489-4a16-895a-5c220bd86f49)
Quick drawing I did

Commit:
create an aggregated itinerary entity

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

Mosaab-Emam committed:
fix: delete duplicate (and outdated) migration file

Migration20251206224129.ts and Migration20251206224359 are effectively duplicates (the latter includes 'cancelled' in the trip_booking status enum.
Having both results in duplicate table errors when attempting to migrate.

jliu1016 committed:
correct departure and arrival time stop time ids

Mosaab-Emam committed:
refactor: adapt to new response shape (wrapped in data: {})

Mosaab-Emam committed:
refactor: ditch the old trips page

Mosaab-Emam committed:
Merge branch 'roundTripKitchenerTOUnion' into nextjs-frontend-setup

Mosaab-Emam committed:
fix: formatting

Commit:
use upsert to prevent race condition

Commit:
merge main

Commit:
start sequence counts from 1

Commit:
use transactional decorator for itinerary creation

Commit:
clean up function names, change to findOrCreate for accuracy

Commit:
use appropriate error object

Commit:
use util function in trip service

Commit:
update dto

Commit:
add automatic trip hashing to itinerary

Commit:
add get existing itineraries

Commit:
update e2e config to match new setup

Commit:
update snap shot json

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/main' into nextjs-frontend-setup

Mosaab-Emam committed:
feat: implementing book round trip page & form (no itinerary creation)

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/feature/65-ItineraryCreation' into nextjs-frontend-setup

Mosaab-Emam committed:
feat: working itinerary creation

Mosaab-Emam committed:
fix: protected -> profile and navbar link

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/feature/116-AggregatedItinerariesView' into nextjs-frontend-setup

Mosaab-Emam committed:
feat: existing itineraries view

Mosaab-Emam committed:
fix: trip hash

Mosaab-Emam committed:
fix: handle date edge case

Mosaab-Emam committed:
fix: protected -> profile

Mosaab-Emam committed:
Merge remote-tracking branch 'origin/feature/itineraryDetails' into nextjs-frontend-setup

Mosaab-Emam committed:
feat: itinerary details dialog
